# About RIMS-seq
Here is a set of programs to analyze data derived from RIMS-seq (Rapid Identification of Methylase Specificity). 
RIMS-seq is a new method to simultaneously sequence bacterial genomes and determine m5C methylase specificities. It uses a simple experimental protocol that closely resembles the DNA-seq protocol for Illumina. For more details on RIMS-seq, please read our preprint (link).

# RIMS-seq pipeline 
## Requirements before starting
Samtools (version xx),
Bedtools (version xx)
and MoSDi (https://bitbucket.org/tobiasmarschall/mosdi/src/master/) are required to be installed on your system. 

## 1. mapping.pl
Adapters need to be removed and the reads mapped to the reference genome. If not using this script, the mapping should be done using BWA-mem.

## 2. split_mapped_reads.pl
This script splits the reads from the bam file generated by BWA. Reads are separated between Read 1 and Read 2 and two files are created accordingly.


## 3. get_motif_all.pl
This script takes the two mpileup files previsouly generated and identifies the methylase specificity using a program called MosDi. 

